File: backend/schemas/organization.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ------------------------------------------------------------- departments

class DepartmentCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    code: Optional[str] = Field(default=None, max_length=50)
    # Only used by super_admin; ignored (forced to caller's college) for admin.
    college_id: Optional[int] = None


class DepartmentUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=150)
    code: Optional[str] = Field(default=None, max_length=50)


class DepartmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    college_id: int
    name: str
    code: Optional[str] = None


# ------------------------------------------------------------------- classes

class ClassCreate(BaseModel):
    department_id: int
    name: str = Field(min_length=1, max_length=150)
    academic_year: Optional[str] = Field(default=None, max_length=20)
    section: Optional[str] = Field(default=None, max_length=20)


class ClassUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=150)
    academic_year: Optional[str] = Field(default=None, max_length=20)
    section: Optional[str] = Field(default=None, max_length=20)


class ClassOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    college_id: int
    department_id: int
    name: str
    academic_year: Optional[str] = None
    section: Optional[str] = None


# ------------------------------------------------------------ staff assignments

class StaffDepartmentAssign(BaseModel):
    department_id: int


class StaffDepartmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    department_id: int
    department_name: str
    is_active: bool
    assigned_at: Optional[datetime] = None


class StaffClassAssign(BaseModel):
    class_id: int
    is_incharge: bool = False


class StaffClassUpdate(BaseModel):
    is_incharge: bool


class StaffClassOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    class_id: int
    class_name: str
    is_incharge: bool
    assigned_at: Optional[datetime] = None

File: frontend/src/types/organization.ts
export interface DepartmentCreate {
  name: string;
  code?: string | null;
  college_id?: number | null;
}
export type DepartmentUpdate = Partial<Pick<DepartmentCreate, "name" | "code">>;
export interface DepartmentOut {
  id: number;
  college_id: number;
  name: string;
  code: string | null;
}

export interface ClassCreate {
  department_id: number;
  name: string;
  academic_year?: string | null;
  section?: string | null;
}
export type ClassUpdate = Partial<Pick<ClassCreate, "name" | "academic_year" | "section">>;
export interface ClassOut {
  id: number;
  college_id: number;
  department_id: number;
  name: string;
  academic_year: string | null;
  section: string | null;
}

export interface StaffDepartmentAssign {
  department_id: number;
}
export interface StaffDepartmentOut {
  id: number;
  department_id: number;
  department_name: string;
  is_active: boolean;
  assigned_at: string | null;
}

export interface StaffClassAssign {
  class_id: number;
  is_incharge?: boolean;
}
export interface StaffClassUpdate {
  is_incharge: boolean;
}
export interface StaffClassOut {
  id: number;
  class_id: number;
  class_name: string;
  is_incharge: boolean;
  assigned_at: string | null;
}
